src/plot_api/plot_api.js

```diff
@@ -2192,6 +2193,181 @@ exports.update = function update(gd, traceUpdate, layoutUpdate, _traces) {
+        }
+    } else {
+        // Either v.dtype was an invalid array type, or this browser doesn't
+        // support this typed array type.
+    }
+    return coercedV;
+}
+
+/**
+ * Recursive helper function to perform decoding
+ */
+function performDecode(v) {
+    if(Lib.isTypedArrayEncoding(v)) {
+        return decodeTypedArray(v);
+    } else if(Array.isArray(v)) {
+        return v.map(performDecode);
+    } else if(Lib.isPlainObject(v)) {
+        var result = {};
+        for(var k in v) {
+            if(v.hasOwnProperty(k)) {
+                result[k] = performDecode(v[k]);
+            }
+        }
+        return result;
+    } else {
+        return v;
+    }
+}
+
+/**
+ * Plotly.decode:
+ * Attempt to recursively decode an object or array into a form supported
+ * by Plotly.js.  This function is the inverse of Plotly.encode.
+ *
+ * @param {object} v: Value to be decoded
+ * @returns {object}: Decoded value
+ */
+exports.decode = function(v) {
+    return performDecode(v);
+};
+
+/**
+ * Get data type string for TypedArray
+ * @param {TypedArray} v: A TypedArray instance
+ * @returns {String}
+ */
+function getDtypeStringForTypedArray(v) {
+    if(typeof Int8Array !== 'undefined' && v instanceof Int8Array) {
+        return 'int8';
+    } else if(typeof Uint8Array !== 'undefined' && v instanceof Uint8Array) {
+        return 'uint8';
+    } else if(typeof Uint8ClampedArray !== 'undefined' && v instanceof Uint8ClampedArray) {
+        return 'uint8_clamped';
+    } else if(typeof Int16Array !== 'undefined' && v instanceof Int16Array) {
+        return 'int16';
+    } else if(typeof Uint16Array !== 'undefined' && v instanceof Uint16Array) {
+        return 'uint16';
+    } else if(typeof Int32Array !== 'undefined' && v instanceof Int32Array) {
+        return 'int32';
+    } else if(typeof Uint32Array !== 'undefined' && v instanceof Uint32Array) {
+        return 'uint32';
+    } else if(typeof Float32Array !== 'undefined' && v instanceof Float32Array) {
+        return 'float32';
+    } else if(typeof Float64Array !== 'undefined' && v instanceof Float64Array) {
+        return 'float64';
+    }
+}
+
+
+/**
+ * Convert a TypedArray instance into a JSON-serializable object that
+ * represents it.
+ *
+ * @param {TypedArray} v: A TypedArray instance
+ *
+ * @returns {object} Object with `dtype` and `value` properties that
+ * represents a TypedArray.
+ */
+function encodeTypedArray(v) {
+    var dtype = getDtypeStringForTypedArray(v);
+    var buffer = b64.encode(v.buffer);
+    return {'value': buffer, 'dtype': dtype};
+}
+
+
+/**
+ * Recursive helper function to perform encoding
+ * @param v
+ */
+function performEncode(v) {
+    if(Lib.isTypedArray(v)) {
+        return encodeTypedArray(v);
+    } else if(Array.isArray(v)) {
+        return v.map(performEncode);
+    } else if(Lib.isPlainObject(v)) {
+        var result = {};
+        for(var k in v) {
+            if(v.hasOwnProperty(k)) {
```

Comment:
Here we may list the keys using [Object.getOwnPropertyNames](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames) and loop through them.